endpoints/data_objects/src/main.cpp

```diff
@@ -1369,12 +1743,35 @@ namespace
 			res.keep_alive(_req.keep_alive());
 
 			try {
+				DataObjInp input{};
+				irods::at_scope_exit free_memory{[&input] { clearKeyVal(&input.condInput); }};
+
+				input.oprType = UNREG_OPR;
+
 				const auto lpath_iter = _args.find("lpath");
 				if (lpath_iter == std::end(_args)) {
 					log::error("{}: Missing [lpath] parameter.", fn);
 					return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
 				}
 
+				std::strncpy(input.objPath, lpath_iter->second.c_str(), sizeof(DataObjInp::objPath));
+				addKeyVal(&input.condInput, COPIES_KW, "0");
```

Comment:
`iunreg` uses `trimUtil` and `rmUtil` to carry out work. The HTTP API sets the minimum number of replicas to keep to 0 and only uses `rcDataObjTrim`.

Everything works, but it made me wonder why `iunreg` relies on `rcDataObjTrim` and `rcDataObjUnlink`.

Reply:
`iunreg` has to be allowed to remove the last replica.  Can the trim endpoint go all the way to zero replicas?

Reply:
if so... we should consider updating iunreg to be safer...

Reply:
I was able to unregister the last replica of a data object with just `rcDataObjTrim`.

Going to have to ask @alanking about `iunreg` to understand why `trim/rmUtil` are necessary.

Reply:
hrm.   surprising.   that is/was not my conceptual model of a trim (which is the sibling of repl).

Reply:
Well, both APIs use the same functionality. That is `dataObjUnlinkS`.

Does `iunreg` allow you to unregister the last replica? The code for that icommand calls the trim API if a replica is explicitly targeted. Otherwise, it calls rmUtil.

Reply:
Hm, right.  So our APIs aren't as cleanly separated as the functions / mental model we're trying to convey.  I think that's where I am getting them crossed up...

`iunreg` should definitely allow for unregistering the last replica.  But we definitely can't do that with `itrim`... I think.

Reply:
I believe that's correct? Certainly worth checking at the bench.

I totally agree there should be a dedicated API endpoint for unregistering replicas. Especially if it should be viewed as its own operation.

Reply:
Playing with `iunreg`, I got the following.
```
$ itouch apple
$ iunreg -N0 -n0 apple # This fails.
remote addresses: X.X.X.X ERROR: trimUtil: trim error for /tempZone/home/rods/apple.  status = -402000 USER_INCOMPATIBLE_PARAMS
$ iunreg apple         # But this succeeds?
```

Why aren't both calls to `iunreg` treated identical?

Reply:
might need more @alanking brain

Reply:
Somehow I think I got the behavior incorrect. The current implementation does not work as expected.

The correct way to do this is to use `rsDataObjTrim` and `rsDataObjUnlink`, just like `iunreg`.

Will tweak.

Reply:
Also, using `RESC_NAME_KW` for `rsDataObjTrim` means the root resource. It cannot be used to target a specific replica within a resource hierarchy.

This makes me think we should only restrict the op to replica nubmers.

Reply:
i think that is *also* the same as iunreg?

Reply:
After some discussion, we're going to make the following changes:
- The trim and remove operations keep the unregister parameter
- The remove operation requires the client to be explicit about the unregister parameter
- The trim operation does not need to enforce the unregister parameter because it can't delete all replicas
- Rename `unregister` parameter to `catalog-only`
- Consider returning `INCOMPATIBLE_PARAMS` when `catalog-only=1` and `no-trash=1` is passed to the remove operation

Reply:
done